src/otx/algo/samplers/balanced_sampler.py

```diff
@@ -0,0 +1,174 @@
+        efficient_mode: bool = False,
+        num_replicas: int = 1,
+        rank: int = 0,
+        drop_last: bool = False,
+        n_repeats: int = 1,
+        generator: torch.Generator | None = None,
+    ):
+        self.dataset = dataset
+        self.samples_per_gpu = samples_per_gpu
+        self.num_replicas = num_replicas
+        self.rank = rank
+        self.drop_last = drop_last
+        self.generator = generator
+        self.repeat = n_repeats
+
+        super().__init__(dataset)
+
+        # img_indices: dict[label: list[idx]]
+        ann_stats = get_idx_list_per_classes(dataset.dm_subset)
+        self.img_indices = {k: v for k, v in ann_stats.items() if len(v) > 0}
+        self.num_cls = len(self.img_indices.keys())
+        self.data_length = len(self.dataset)
+        self.num_trials = int(self.data_length / self.num_cls)
+
+        if efficient_mode:
+            # Reduce the # of sampling (sampling data for a single epoch)
+            num_tail = min(len(cls_indices) for cls_indices in self.img_indices.values())
+            if num_tail > 1:
+                base = 1 - (1 / num_tail)
+                num_reduced_trials = int(math.log(0.001, base))
+                self.num_trials = min(num_reduced_trials, self.num_trials)
+
+        self.num_samples = self._calculate_num_samples()
+
+    def _calculate_num_samples(self) -> int:
+        num_samples = self.num_trials * self.num_cls * self.repeat
+
+        if self.num_replicas > 1:
+            # If the dataset length is evenly divisible by # of replicas, then there
+            # is no need to drop any data, since the dataset will be split equally.
+            if self.drop_last and num_samples % self.num_replicas != 0:
+                # Split to nearest available length that is evenly divisible.
+                # This is to ensure each rank receives the same amount of data when
+                # using this Sampler.
+                num_samples = math.ceil(
+                    # `type:ignore` is required because Dataset cannot provide a default __len__
+                    # see NOTE in pytorch/torch/utils/data/sampler.py
+                    (num_samples - self.num_replicas) / self.num_replicas,
+                )
+            else:
+                num_samples = math.ceil(num_samples / self.num_replicas)
+            self.total_size = num_samples * self.num_replicas
+
+        return num_samples
+
+    def __iter__(self):
+        """Iter."""
+        if self.generator is None:
+            seed = int(torch.empty((), dtype=torch.int64).random_().item())
+            generator = torch.Generator()
+            generator.manual_seed(seed)
+        else:
+            generator = self.generator
+
+        indices = []
+        for _ in range(self.repeat):
+            for _ in range(self.num_trials):
+                indice = torch.cat(
+                    [
+                        torch.tensor(self.img_indices[cls_indices], dtype=torch.int64)[
```

Comment:
`self.img_indices[cls_indices]`  can be a large vector. Change this to `torch.tensor` in advance.

Reply:
I modified this so that `img_indices` has a value as a tensor when it is created.
- https://github.com/openvinotoolkit/training_extensions/pull/2993/commits/6eb345a7526d42dafb6351fc7de10e41341786b8